programs/survival/bucketstack.sc

```diff
@@ -0,0 +1,129 @@
+		l('pufferfish',1),
+		l('salmon',1),
+		l('cod',1),
+		l('tropical_fish',1),
+		l('axolotl',1)
+	);
+	return(
+		m(
+			l('scope','player'),
+			l('stay_loaded',true)
+		)
+	);
+);
+
+__on_player_collides_with_entity(player,entity) ->
+(
+	if(entity~'pickup_delay' == 0,
+		l(item,count,nbt) = entity~'item';
+		if(item~'_bucket',
+			//if it's an item we're picking up that is a non-empty bucket
+			//first determine the max stack size
+			type = null;
+			while(type == null,length(global_buckets),
+				type = item~(keys(global_buckets):_);
+			);
+			max_stack = global_buckets:type;
+			//then find a non-full slot to put it in
+			//modified lines 103 and 104 of gnembon's shulkerboxes.sc to find a non-full slot
+			slot = -1;
+			while(count && (slot = inventory_find(player,item,slot+1)) != null && slot < 36,inventory_size(player)-5,
+				//using copy() because script kept throwing errors here
+				l(slot_item,slot_count,slot_nbt) = copy(inventory_get(player,slot));
+				if(slot_count < max_stack,
+					//attempt to pick up items
+					count = __pickup(player,entity,item,slot,count,slot_count,max_stack);
+				);
+			);
+			//if all possible stacks are full, find as many empty slots as necessary
+			//slot < 36 and inventory_size(player)-5 unless you want to wear the buckets
+			while(count && (slot = inventory_find(player,null)) != null && slot < 36,inventory_size(player)-5,
+				count = __pickup(player,entity,item,slot,count,0,max_stack);
+			);
+		);
+	);
+);
+
+__pickup(player,entity,item,slot,count,slot_count,max_stack) ->
+(
+	//figure out how many items to move
+	//and move them
+	move = min(max_stack - slot_count,count);
+	inventory_set(player,slot,slot_count + move,item);
+	count = count - move;
+	//if all possible slots are full, leave whatever is left on the ground
+	//modified line 90 of gnembon's shulkerboxes.sc to change the number of items in the item entity
+	modify(entity,'nbt_merge','{Item:{Count:' + count + 'b}}');
+	//line 68 of gnembon's shulkerboxes.sc to play the item pickup sound
+	if(move,
+		sound('entity.item.pickup',pos(player),0.2,(rand(1)-rand(1))*1.4+2.0, 'player');
+	);
+	return(count);
+);
+
+//WRITTEN BY FIRIGION
+
+__on_player_right_clicks_block(player,item_tuple,hand,block,face,hitvec) ->
+(
+	if(hand == 'mainhand',
+		slot = player~'selected_slot',
+		//-1 is offhand
+		slot = -1;
+	);
+	if(item_tuple:0~'_bucket' && !item_tuple:0~'milk',
+		schedule(0,'test_bucket_used',player,hand,item_tuple,slot);
+	);
+);
+
+//only for milk
+__on_player_finishes_using_item(player,item_tuple,hand) ->
+(
+	if(hand == 'mainhand',
+		slot = player~'selected_slot',
+		//-1 is offhand
+		slot = -1;
+	);
+  	if(item_tuple:0~'_bucket' && item_tuple:0~'milk',
+		schedule(0,'test_bucket_used',player,hand,item_tuple,slot);
+	);
+);
+
+test_bucket_used(player, hand, item_tuple, slot) ->
+(
+	l(item,count,nbt) = item_tuple;
+	//trigger only if bucket was used
+ 	if(inventory_get(player,slot):0 == 'bucket' || item~'milk',

```

Comment:
Don't we already know that the player has bucket by now?

Reply:
We previously checked that the item that was being used was a bucket variant. At this point, the bucket has been used and the slot _should_ still have an empty bucket or nonzero milk buckets. In the event that a player is swapping hands quickly or using Tweakeroo to rapidly change the item in a slot, we recheck before setting a bucket to prevent bucket duping.